### ISSUE TITLE
Should all parameters in Generic be enforced distinct?

### ISSUE THREAD

**Issue by ilevkivskyi** [ambv/typehinting#134]
Currently PEP says: Each type variable argument to Generic must be distinct. This is
thus invalid  `class Pair(Generic[T, T]):  ...`

However, this is not enforced, one could create such class without any warning. Is this a bug?
If yes, one could add following to `__getitem__` in `GenericMeta` at the end of block considering `self.__parameters__ is None`:
```python
if len(set(params)) != len(params):
    raise TypeError("All type variables in Generic[...] must be distinct.")
```

There is a related question: Should be two `TypeVar`s with the same name, constraints, etc. considered equal? For example, if I have:
```python
T = TypeVar('T')
T1 = TypeVar('T')
```
then is it allowed to write `class Pair(Generic[T1, T]):  ...`? If no, then one could add something like
```python
def __hash__(self):
    return hash((self.__name__, self.__covariant__,
                         self.__contravariant__, self.__bound__,
                         self.__constraints__))

def __eq__(self, other):
    if not isinstance(other, TypeVar):
        return NotImplemented
    return (self.__name__ == other.__name__ and
                self.__covariant__ == other.__covariant__ and
                self.__contravariant__ == other.__contravariant__ and
                self.__bound__ == other.__bound__ and
                self.__constraints__ == other.__constraints__)
```
or similar to `TypeVar`.

**Comment by ilevkivskyi** [comment]
I have found the following in GENERICS.rst:
"""
Note that the argument to TypeVar() must be a string, and it must be assigned to a variable with exactly that name. Type variables cannot be redefined in the same module. These constraints are enforced by the type checker (but not by the runtime implementation of TypeVar())
"""
I think it is important, since it means that the example above with `T1 = TypeVar('T')` is not allowed. Maybe add it to the PEP?

**Comment by gvanrossum** [comment]
The first thing (Generic[T,T] is invalid) is a check that I plan to fix eventually. The PEP already mentions this.

The second thing (T1 = TypeVar('T') is invalid) is unrealistic to check for. It should be added to the PEP.

**Comment by gvanrossum** [comment]
I've added words to the PEP describing restrictions on type variables (rev 71bf400). The check against Generic[T,T] is still to do.

**Comment by ilevkivskyi** [comment]
I don't think it requires a separate issue, but currently `typing.py` contains this
```python
class Mapping(Sized, Iterable[KT], Container[KT], Generic[KT, VT_co],
              extra=collections_abc.Mapping):
    pass
```
However, the logic is not clear. The `Mapping` should be already generic in `KT` by inheriting from generics `Iterable` and `Container`. Should not we only need to write `Generic[VT_co]` in the definition of `Mapping`?
I think that in general things like
```python
class MyClass(MyGenClass[T], Generic[T]): ...
```
should be prohibited, since logically it is similar to `Generic[T, T]`

**Comment by gvanrossum** [comment]
I think you're right in that using just Generic[VT_co] in the declaration of Mapping is sufficient. But I don't think it's wrong to specify both -- redundant type parameters in different base classes should be resolved amicably (as can be already seen by the reuse of Kt for Iterable and Container).

**Comment by ilevkivskyi** [comment]
As I understand, using `KT` in both `Iterable` and `Container` bases provides some useful type information. For example, if I write
```python
class MyStrangeClass(Iterable[T1], Container[T2]):
    ...
```
then later I could write `MyStrangeClass[int, str]`. At the same time if I use the same variable, then I can only write `MyStrangeClass[int]`.
In any case, I agree that some redundant type info is OK. But I think that the presence of `Generic[KT, VT_co]` in bases of `Mapping` could confuse a reader of `typing.py` and could make an impression that such (a bit verbose) syntax is obligatory. In my opinion it would be much clearer for the reader if there be `Generic[VT_co]` instead.

**Comment by ilevkivskyi** [comment]
This could be now closed since PR #144 fixed this.
